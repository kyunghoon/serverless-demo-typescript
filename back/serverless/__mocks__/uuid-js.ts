const uuid = jest.genMockFromModule('uuid-js');

uuid.__value = '00000000-0000-0000-0000-000000000000';

uuid.create = jest.fn(() => uuid.__value);

uuid.__set = (value) => {
  uuid.__value = value;
};

module.exports = uuid;
