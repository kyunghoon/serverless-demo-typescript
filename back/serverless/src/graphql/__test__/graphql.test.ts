 jest.mock('../collections/auth');

const createUser = async (name, id) => {
  const graphql = require('..').default;
  require('uuid-js').__set(`id-${id}`);
  const query = `
    mutation {
      createUser (email: "${name}", password: "passpass") {
        id name email token screenName permissions
      }
    }`;
  return graphql(query, {});
};

const expectOk = result => {
  expect(result.errors).toBeUndefined();
  expect(result.data).toBeDefined();
};

const expectError = result => {
  expect(result.errors).toBeDefined();
};

describe('graphql', () => {
  it('should get the viewer', async () => {
    const graphql = require('..').default;
    const query = `
      query {
        viewer {
          id name email token screenName permissions
        }
      }`;
    const result = await graphql(query, {});
    expectOk(result);
    expect(result).toMatchSnapshot();
  });

  it('should update a user', async () => {
    const graphql = require('..').default;
    const result = await createUser('update@test.com', '0');
    expectOk(result);
    const query = `
      mutation {
        updateUser (id: "${result.data.createUser.id}", name: "updated name", screenName: "updated screenname") {
          id name email token screenName permissions
        }
      }`;
    const result2 = await graphql(query, {});
    expectOk(result2);
    expect(result2).toMatchSnapshot();
  });

  it('should fail to update a non-existing user', async () => {
    const graphql = require('..').default;
    const query = `
      mutation {
        updateUser (id: "1003", name: "updated name", screenName: "updated screenName") {
          id name email token screenName permissions
        }
      }`;
    const result = await graphql(query, {});
    expectError(result);
    expect(result).toMatchSnapshot();
  });

  it('should create a user', async () => {
    const result = await createUser('create@test.com', '1');
    expectOk(result);
    expect(result).toMatchSnapshot();
  });

  it('should fail to create a user with an existing email', async () => {
    const result = await createUser('existing@test.com', '2');
    expectOk(result);
    const result2 = await createUser('existing@test.com', '3');
    expectError(result2);
    expect(result2).toMatchSnapshot();
  });

  it('should remove a user', async () => {
    const graphql = require('..').default;
    const result = await createUser('remove@test.com', '4');
    expectOk(result);
    const query = `
      mutation {
        removeUser (id: "${result.data.createUser.id}") {
          id
        }
      }`;
    const result2 = await graphql(query, {});
    expectOk(result2);
    expect(result2).toMatchSnapshot();
  });

  describe('login', () => {
    const graphql = require('..').default;
    let user = null;

    beforeAll(async () => {
      const result = await createUser('login@test.com', '5');
      expectOk(result);
      user = result.data.createUser;
    });

    it('should find a user', async () => {
      const query = `
        query {
          user (id: "${user.id}") {
            id name email token screenName permissions
          }
        }`;
      const result = await graphql(query, {});
      expectOk(result);
      expect(result).toMatchSnapshot();
    });

    it('should login with a correct password', async () => {
      const query = `
        query {
          login (email: "${user.email}", password: "passpass") {
            id name email token screenName permissions
          }
        }`;
      const result = await graphql(query, {});
      expectOk(result);
      expect(result).toMatchSnapshot();
    });

    it('should not login with a bad password', async () => {
      const query = `
        query {
          login (email: "${user.email}", password: "badbad") {
            id name email token screenName permissions
          }
        }`;
      const result = await graphql(query, {});
      expectError(result);
      expect(result).toMatchSnapshot();
    });
  });
});
