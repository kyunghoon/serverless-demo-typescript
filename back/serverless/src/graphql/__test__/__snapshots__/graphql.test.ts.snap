exports[`graphql login should find a user 1`] = `
Object {
  "data": Object {
    "user": Object {
      "email": "login@test.com",
      "id": "id-5",
      "name": null,
      "permissions": Array [
        "UPDATE_USER",
        "DELETE_USER",
      ],
      "screenName": "login",
      "token": null,
    },
  },
}
`;

exports[`graphql login should login with a correct password 1`] = `
Object {
  "data": Object {
    "login": Object {
      "email": "login@test.com",
      "id": "id-5",
      "name": null,
      "permissions": Array [
        "UPDATE_USER",
        "DELETE_USER",
      ],
      "screenName": "login",
      "token": null,
    },
  },
}
`;

exports[`graphql login should not login with a bad password 1`] = `
Object {
  "data": Object {
    "login": null,
  },
  "errors": Array [
    [GraphQLError: invalid password],
  ],
}
`;

exports[`graphql should create a user 1`] = `
Object {
  "data": Object {
    "createUser": Object {
      "email": "create@test.com",
      "id": "id-1",
      "name": null,
      "permissions": Array [
        "UPDATE_USER",
        "DELETE_USER",
      ],
      "screenName": "create",
      "token": null,
    },
  },
}
`;

exports[`graphql should fail to create a user with an existing email 1`] = `
Object {
  "data": Object {
    "createUser": null,
  },
  "errors": Array [
    [GraphQLError: email already in use],
  ],
}
`;

exports[`graphql should fail to update a non-existing user 1`] = `
Object {
  "data": Object {
    "updateUser": null,
  },
  "errors": Array [
    [GraphQLError: The conditional request failed],
  ],
}
`;

exports[`graphql should get the viewer 1`] = `
Object {
  "data": Object {
    "viewer": Object {
      "email": "guest@guests.com",
      "id": "1",
      "name": "guest",
      "permissions": Array [
        "UPDATE_USER",
        "DELETE_USER",
      ],
      "screenName": "guest",
      "token": null,
    },
  },
}
`;

exports[`graphql should remove a user 1`] = `
Object {
  "data": Object {
    "removeUser": Object {
      "id": "id-4",
    },
  },
}
`;

exports[`graphql should update a user 1`] = `
Object {
  "data": Object {
    "updateUser": Object {
      "email": "update@test.com",
      "id": "id-0",
      "name": "updated name",
      "permissions": Array [
        "UPDATE_USER",
        "DELETE_USER",
      ],
      "screenName": "updated screenname",
      "token": null,
    },
  },
}
`;
